File: examples/code/integrations/oauth2/custom_auth.js
import { Arcade } from "@arcadeai/arcadejs";

const client = new Arcade(); // Automatically finds the `ARCADE_API_KEY` env variable

const userId = "user@example.com";

// Start the authorization process
let authResponse = await client.auth.start(userId, "hooli", [
  "scope1",
  "scope2",
]);

if (authResponse.status !== "completed") {
  console.log("Please complete the authorization challenge in your browser:");
  console.log(authResponse.url);
}

// Wait for the authorization to complete
authResponse = await client.auth.waitForCompletion(authResponse);

// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
const token = authResponse.context.token;
// TODO: Do something interesting with the token...

File: examples/code/custom_auth.py
from arcade.client import Arcade
from arcade.client.schema import AuthProvider
from arcade.client.errors import APITimeoutError

client = Arcade(api_key="your_api_key")

user_id = "user@example.com"

# Start the authorization process
auth_response = client.auth.authorize(
    provider=AuthProvider.google,
    scopes=["https://www.googleapis.com/auth/gmail.readonly"],
    user_id=user_id,
)

if auth_response.status != "completed":
    print("Please complete the authorization challenge in your browser:")
    print(auth_response.auth_url)

while auth_response.status != "completed":
    try:
        auth_response = client.auth.status(auth_response, wait=60)
    except APITimeoutError:
        continue